doc/README.twom.md

```diff
@@ -0,0 +1,189 @@
+
+This means that so long as the final header write is atomic, transactions are always completely safe.
+
+## Twoskip downsides:
+
+- Three fsyncs per transaction is kinda expensive!
+- Every write causes the file to be extended, as twoskip format uses exact file size and doesn’t allocate slop.
+- A single write does amplify into 1+N (the level of the record) writes; the append of the new record, and the updates of the immediately previous records in the skiplist at each level.  Since half the records are level 1, half that level 2, etc - the average write updates three separate previous records.  In a recently repacked file, the previous records are likely to be adjacent or nearly so for the lower levels, but this becomes less so over time, so repacks are valuable for both data locality and size reduction.
+- The repack operation requires the database to be remain exclusively locked for the duration, which is slow on large files (particularly along with the above lack of slop).
+
+## The Twom format
+
+Why the name?  It’s twoskip, but with mmap and MVCC.  Two Ms.  And the “tombstone” records, rather than just causing the skiplist to be rewritten without the deleted record in it, remain in the linked lists right until a repack, with an ANCESTOR pointer back to the previous entity with that key.
+
+This tombstone allows the MVCC transaction, because you can walk the current lowest level linked list, and any time you find a record with an offset in the future of the file size at the time you started the transaction, walk back up the ancestor chain (which is a single linked list backward through the file, to the original ADD record which was the first time the key was seen in this file).
+
+With MVCC transactions, we can then re-implement the repack operation; instead of having to lock the file for the entire repack, we can take a read-only lock and yield it many times while walking the original records and adding them to the new file in sorted order, and then (still releasing locks as many times as we like) read the remainder of the file and replay the operations (adds, replaces and deletes) to the new file.  This doesn’t require any exclusive lock at any point, so readers can continue throughout the repack, even until the rename - and either get the old file or the new file.
+
+Remaining issues:
+
+- We still use 3 fsyncs - there’s no way around this without risking a crash leaving a file in a broken state.  We could theoretically still get a clean read after a single failure by checking for end pointers during the “locate” phase and falling back to the lowest level pointers - but if we then started another write; there would be invalid pointers somewhere in the file and it would corrupt over time.  Nope.
+- The write amplification issue is still the same, twom is basically the same format as twoskip, just with the addition of ancestor pointers and richer tombstones.
+
+### Navigating the code - data structures:
```

Comment:
Both "Navigating the code" sections might better go in the C file?

Reply:
There's a bunch of comments in the C file too!

Reply:
OK. You still might want to move both of the "Navigating the code" section at the of the README, so that that the Locking section isn't interleaved.